File: inst/python/extract_variants.py
#!/usr/bin/env python

# Import modules
import argparse
import vcf
import os.path
import re

# Argument parser
parser = argparse.ArgumentParser(epilog='Extracts variant data from a VCF.')
parser.add_argument('input', type=str, help='input VCF file path')
parser.add_argument('sample', type=str, help='sample to extract')
parser.add_argument('output', type=str, help='output file path')
parser.add_argument('-f', '--filter-depth', type=int, dest='filter_depth',
        default=10, help='filter variants with depth below <filter>')
args = parser.parse_args()

# Sample to extract from VCF
sample = args.sample

# Remove output file if already existing
if os.path.isfile(args.output):
    os.remove(args.output)

# Open output file for appending
output_file = open(args.output, 'a')

# Header row
header = 'chr\t' + \
    'pos\t' + \
    'rsID\t' + \
    'gene\t' + \
    'ENSGID\t' + \
    'ENSTID\t' + \
    'REF\t' + \
    'ALT\t' + \
    'impact\t' + \
    'effect\t' + \
    'feature\t' + \
    'biotype\t' + \
    'DP\t' + \
    'AD1\t' + \
    'AD2\t' + \
    'A1\t' + \
    'A2\t' + \
    'warnings\n'

# Write header to file
output_file.write(header)

# Initialise set for unique lines
unique_lines = set()

# Function for ordering on columns
def col_sort(string):
    s = string.split('\t')
    return [ s[0], int(s[1]), s[3].upper(), s[4].upper(),
            s[5].upper().replace('.', ':'),
            s[8].upper(), s[9].upper(), s[10].upper() ]

# Open input VCF file
vcf_reader = vcf.Reader(open(args.input, 'r'))

for record in vcf_reader:
    ref=record.REF
    alt=str(record.ALT).strip('[]')
    id=record.ID
    chrom=record.CHROM
    pos=record.POS
    
    # Skip non-SNVs
    if len(ref) > 1 or len(alt) > 1:
        continue

    # Collect genotype info (skip record if any is missing)
    try:
        gt = record.genotype(str(sample))['GT']
        ad = record.genotype(str(sample))['AD']
        dp = record.genotype(str(sample))['DP']
    except AttributeError:
        continue
    
    # Collect annotation infor (skip record if missing)
    try:
        ann=record.INFO['ANN']
        qual=record.QUAL
        filt=record.FILTER
    except KeyError:
        continue

    # Skip variant if below filtering depth
    if dp < args.filter_depth:
        continue

    # Get filter info
    if filt:
         filt=filt[0]
    else:
         filt='None'

    # Skip record if it's not passing filters
    if filt != 'None':
        continue

    # Make AD into list if only one value is available
    try:
        ad[0] = ad[0]
    except TypeError:
        ad = [ad, 0]

    if gt:

        gts = gt.split('/')
        A1 = gts[0]
        A2 = gts[1]

        if A1=='0':
            A1GT=ref
        else:
            if A1=='1':
                A1GT=alt[0]
            else:
                if A1=='2':
                    A1GT=alt[1]
                else:
                    A1GT='NONE'

        if A2=='0':
           A2GT=ref
        else:
            if A2=='1':
                A2GT=alt[0]
            else:
                if A2=='2':
                    A2GT=alt[1]
                else:
                    A2GT='NONE'

        result = ['%s %s %s %s %s %s %s %s %s %s' %
            (line.split('|')[3], line.split('|')[4], line.split('|')[2],
             line.split('|')[1], line.split('|')[5], line.split('|')[6],
             line.split('|')[7], line.split('|')[9], line.split('|')[10],
             line.split('|')[15]) for line in ann]

        result = set(result)
        priority_list = list(reversed(['HIGH', 'MODERATE', 'LOW', 'MODIFIER']))
        max_index = -1

        for line in result:
            gene, ensgid, impact, effect, feature, enst, biotype, nucl, aacid, warnings = line.split(' ')
            max_index = max(max_index, priority_list.index(impact))
        max_impact = priority_list[max_index]

        for line in result:
            gene, ensgid, impact, effect, feature, enst, biotype, nucl, aacid, warnings = line.split(' ')
            if impact == max_impact:
                current = str(chrom) + "\t" + \
                          str(pos) + "\t" + \
                          str(id) + "\t" +  \
                          str(gene) + "\t" + \
                          str(ensgid) + "\t" + \
                          str(enst) + "\t" + \
                          str(ref) + "\t" + \
                          str(alt[0]) + "\t" + \
                          str(impact) + "\t" + \
                          str(effect) + "\t" + \
                          str(feature) + "\t" + \
                          str(biotype) + "\t" + \
                          str(dp) + "\t" + \
                          str(ad[0]) + "\t" + \
                          str(ad[1]) + "\t" + \
                          str(A1GT) + "\t" + \
                          str(A2GT) + "\t" + \
                          str(warnings) + "\n"

                # Add to lines (if unique)
                if current not in unique_lines:
                    unique_lines.add(current)

# Sort and write to file
output_file.writelines(sorted(unique_lines, key=col_sort))

# Close output file
output_file.close()

File: R/extract_variants.R
#' Extract relevant variant information from a VCF file
#'
#' This function is a wrapper to the "extract_variants.py" Python script, which
#' extracts relevant variant information from a VCF file. Extraction is
#' performed to facilitate and accelerate the cell authentication procedures,
#' which is especially relevant when more than one pairwise comparison will be
#' performed on the same sample.
#'
#' @param vcf_file The VCF file from which variants will be extracted.
#' @param sample The sample in the VCF that will be extracted.
#' @param output_file Results will be output to this file
#' @return Does not return any data object, but output results to output_file
#' @examples
#' extract_variants_file(vcf_file, sample, output_file, filter_depth=10,
#'                       python=FALSE)

#' @export
#' @rdname extract_variants
extract_variants = function(vcf_file,
                            sample,
                            output_file,
                            filter_depth=10,
                            python=FALSE) {
    
    # Extract variants
    if (python) {

        # Use Python
        message('[extracting with Python]')

        # Source the Python script
        command = system.file('python/extract_variants.py',
							  package='CellAuthentication')

        # Run Python code
        system2(command, args=c(vcf_file, sample, output_file, '-f',
				filter_depth))

    } else {

        # Use R
        message('[extracting with R]')

        # Read VCF file
        vcf = VariantAnnotation::readVcf(vcf_file)

        # Gather relevant information to data GRanges object
        gr = SummarizedExperiment::rowRanges(vcf)
        gr$ANN = VariantAnnotation::info(vcf)$ANN
        gr$DP = as.data.frame(VariantAnnotation::geno(vcf)$DP)[[sample]]
        gr$AD = as.data.frame(VariantAnnotation::geno(vcf)$AD)[[sample]]
        gr$GT = as.data.frame(VariantAnnotation::geno(vcf)$GT)[[sample]]

        # Set ALT as character
        gr$ALT = S4Vectors::unstrsplit(IRanges::CharacterList(gr$ALT))

        # Remove variants not passing variant calling filters
        gr = gr[gr$FILTER == 'PASS', ]
        gr$FILTER = NULL

        # Remove variants below the given depth threshold
        gr = gr[gr$DP >= filter_depth & !is.na(gr$DP), ]

        # Convert to data frame
        data = GenomicRanges::as.data.frame(gr)

        # Remove non-SNVs
        data = data[nchar(data$REF) == 1 &
                    nchar(data$ALT) == 1, ]

        # Get rsIDs if existing
        data$rsID = row.names(data)
        data[!grepl('^rs[0-9]+', data$rsID), 'rsID'] = 'None'

        # Remove unwanted columns
        row.names(data) = NULL
        data$end = NULL
        data$width = NULL
        data$strand = NULL
        data$paramRangeID = NULL
        data$QUAL = NULL

        # Separate allelic depths
        data = tidyr::separate(data=data, col=AD, sep='\\,\\ ',
                               into=c('AD1', 'AD2'), remove=TRUE)
		data$AD1 = gsub('c\\(', '', data$AD1)
		data$AD2 = gsub('\\)', '', data$AD2)

        # Add alleles
        data = tidyr::separate(data=data, col=GT, sep='/', into=c('A1', 'A2'),
                        remove=TRUE)
        data[data$A1 == 0, 'A1'] = data[data$A1 == 0, 'REF']
        data[data$A1 == 1, 'A1'] = data[data$A1 == 1, 'ALT']
        data[data$A2 == 0, 'A2'] = data[data$A2 == 0, 'REF']
        data[data$A2 == 1, 'A2'] = data[data$A2 == 1, 'ALT']

		# Initialise empty data frame for final results
		results = data.frame(effect=character(),
			impact=character(),
			gene=character(),
			ENSGID=character(),
			feature=character(),
			ENSTID=character(),
			biotype=character(),
			warnings=character(),
			seqnames=integer(),
			start=integer(),
			rsID=character(),
			REF=character(),
			ALT=character(),
			DP=integer(),
			AD1=integer(),
			AD2=integer(),
			A1=character(),
			A2=character(), stringsAsFactors=FALSE)

		# Loop over each SNV
		for (n in c(1:nrow(data))) {

			# Get annotation data for current SNV
			ann = data[n, 'ANN'][[1]]

			# Separate into columns
			ann = tidyr::separate(as.data.frame(ann), col=ann, sep='\\|',
				into=c('ALT', 'effect', 'impact', 'gene', 'ENSGID', 'feature',
					   'ENSTID', 'biotype', 'rank', 'HGSV.c', 'HGSV.p',
					   'cDNA.pos', 'CDS.pos', 'protein.pos', 'distance',
					   'warnings'), remove=TRUE)

			# Remove unwanted data columns
			ann$ALT = NULL
			ann$rank = NULL
			ann$HGSV.c = NULL
			ann$HGSV.p = NULL
			ann$cDNA.pos = NULL
			ann$CDS.pos = NULL
			ann$protein.pos = NULL
			ann$distance = NULL

			# Keep only the highest impact SNV(s)
			impacts = unique(ann$impact)
			if ('HIGH' %in% impacts) {

				ann = ann[ann$impact == 'HIGH', ]

			} else if ('MODERATE' %in% impacts) {

				ann = ann[ann$impact == 'MODERATE', ]

			} else if ('LOW' %in% impacts) {

				ann = ann[ann$impact == 'LOW', ]

			}

			# SNV data columns
			data.cols = c('seqnames', 'start', 'rsID', 'REF', 'ALT', 'DP',
						  'AD1', 'AD2', 'A1', 'A2')

			# Add SNV data to each annotation
			for (col in data.cols) {
				ann[[col]] = data[n, col]
			}

			# Append to final results data frame
			results = rbind(results, ann)

		}

		# Finalise output
		results = results[c('seqnames', 'start', 'rsID', 'gene', 'ENSGID', 
                            'ENSTID', 'REF', 'ALT', 'impact', 'effect',
							'feature', 'biotype', 'DP', 'AD1', 'AD2', 'A1',
							'A2', 'warnings')]

		names(results) = c('chr', 'pos', names(results)[3:18])

        # Remove duplicate rows (if present)
        results = unique(results)

        # Sort output
        results = results[order(results$chr, results$pos,
                                results$ENSGID, results$ENSTID), ]

		# Write results to file
		write.table(results, output_file, sep='\t', row.names=FALSE, 
                    quote=FALSE)
    }
}
